### PR TITLE
Add future activity digest for slack

### LINKED ISSUE #77: Report future activities in Slack channel
Provide a function which will list all activities happening over the next x time, in hours. 
Only send to slack if there are actual activities to list, no empty or "informative" posts.
Rate limit?
Base on Community, which for today is only one event

### PR DESCRIPTION
**Pull request by brettski**
Add future activity digest for slack. Much refactoring too
Add's new graph function under community which posts a digest of future activities coming to THAT.us. 

The function takes `hours`, The floor value of the current time is used and sessions are queried based on that time through the number of hours provided. 

fixes #77

### CODE REVIEW

**Comment by brettski** [comment]
Scope added to Auth0 and assigned appropriately. Admin and bot rolls have scope. This has been tested locally (query in insomnia)

This is good to go, the last piece on getting to work on a cadence is setting up the auth scheduler (external to this repo).

File: src/graphql/resolvers/queries/event.js [hunk 1]
```diff
@@ -51,18 +51,18 @@ export const fieldResolvers = {
...
       { dataSources: { firestore } },
     ) => {
       dlog('sessions');
 
-      return sessionStore(firestore).findAllApprovedByEventIdAtDate(
-        id,
-        onOrAfter,
-        daysAfter,
-      );
+      const result = await sessionStore(
```
Comment (csell5) [inline src/graphql/resolvers/queries/event.js:61]: total nit.. but if we want to keep things similar then just chain a .then like we did above and drop the async/await. end result is the same but jsut consistent.

File: src/graphql/resolvers/queries/communities.js [hunk 2]
```diff
@@ -60,5 +61,50 @@ export const fieldResolvers = {
...
+      const sessionFuncs = activeEvents.map(ev =>
+        sessionStore(firestore).findAllApprovedByEventIdAtDateHours(
+          ev.id,
+          atDate,
+          hoursAfter,
+        ),
+      );
+      const sessionRefs = await Promise.all(sessionFuncs);
+      const sessions = [];
+      sessionRefs.forEach(s => sessions.push(...s));
```
Comment (csell5) [inline src/graphql/resolvers/queries/communities.js:101]: why not just use .map? and even having said that.. not sure I get why take one and just push into another?

File: src/graphql/resolvers/queries/communities.js [hunk 3]
```diff
@@ -60,5 +61,50 @@ export const fieldResolvers = {
...
+      if (start) digestStart = start;
+
+      let atDate;
+      // Date as of now min, sec, ms set to zero
+      if (digestStart === 'CURRENT_HOUR') {
+        atDate = new Date(new Date(Date.now()).setMinutes(0, 0, 0));
+      } else if (digestStart === 'NEXT_HOUR') {
+        // now + 1 hour (3600000 ms)
+        atDate = new Date(
+          new Date(Date.now()).setMinutes(0, 0, 0) + 60 * 60 * 1000,
```
Comment (csell5) [inline src/graphql/resolvers/queries/communities.js:86]: I pretty sure `new Date()` is the same as `new Date(Date.now())`

File: src/graphql/resolvers/queries/communities.js [hunk 4]
```diff
@@ -60,5 +61,50 @@ export const fieldResolvers = {
...
+      const activeEvents = await communityStore(firestore).findIsActiveEvents(
+        name,
+      );
+      let digestStart = 'CURRENT_HOUR';
+      if (start) digestStart = start;
+
+      let atDate;
+      // Date as of now min, sec, ms set to zero
+      if (digestStart === 'CURRENT_HOUR') {
+        atDate = new Date(new Date(Date.now()).setMinutes(0, 0, 0));
```
Comment (csell5) [inline src/graphql/resolvers/queries/communities.js:82]: see comment below

File: src/graphql/resolvers/queries/event.js (hunk 1 above)
Comment (brettski) [reply src/graphql/resolvers/queries/event.js:61]: I agree with this nit

File: src/lib/slack/callSlackHook.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    process.env.TEST_SLACK_NOTIFICATIONS
+  ) {
+    const slackUrl = envConfig.slackWebhookUrl;
+    fetch(slackUrl, {
+      method: 'post',
+      body: JSON.stringify(hookBody),
+      headers: { 'Content-Type': 'application/json' },
+    })
+      .then(res => res.text())
+      .then(res => dlog('Slack webhood response: %o', res))
```
Comment (csell5) [inline src/lib/slack/callSlackHook.js:20]: did you want to return the result from the function? cause it's being swallowed up by the next .then.

File: src/graphql/resolvers/queries/communities.js (hunk 2 above)
Comment (brettski) [reply src/graphql/resolvers/queries/communities.js:101]: Map will return a single array in this case? I am taking 1 + N array of objects and making them into one array of objects.

File: src/graphql/resolvers/queries/communities.js (hunk 3 above)
Comment (brettski) [reply src/graphql/resolvers/queries/communities.js:86]: I just tested in REPL, it is not the same.

File: src/graphql/resolvers/queries/communities.js (hunk 4 above)
Comment (brettski) [reply src/graphql/resolvers/queries/communities.js:82]: Tested in REPL, it is not the same

**Comment by csell5** [comment]
> Map will return a single array in this case? I am taking 1 + N array of objects and making them into one array of objects.  

ahhh so s is also an array?

File: src/lib/slack/callSlackHook.js (hunk 5 above)
Comment (brettski) [reply src/lib/slack/callSlackHook.js:20]: Only returning the result during dev, not need response. Perhaps should send an error to sentry other than nowhere. Wasn't to concerned if this failed.

**Comment by brettski** [comment]
It is taking the result from the Promise.all() and combining them all together.

Brett


On September 3, 2020 at 8:45 PM, Clark Sell (notifications@github.com) wrote:
Map will return a single array in this case? I am taking 1 + N array of objects and making them into one array of objects.
ahhh so s is also an array?

**Comment by csell5** [comment]
> Tested in REPL, it is not the same

??? I just did and they were... 

const x = new Date();
const y = new Date(Date.now());

what are we missing?

**Comment by brettski** [comment]
I wasn't in the REPL!!! 🤣



On September 3, 2020 at 8:50 PM, Clark Sell (notifications@github.com) wrote:
Tested in REPL, it is not the same
??? I just did and they were...
const x = new Date();
const y = new Date(Date.now());
what are we missing?

File: src/graphql/resolvers/queries/communities.js (hunk 3 above)
Comment (csell5) [reply src/graphql/resolvers/queries/communities.js:86]: so I did this

```
const x = new Date();
const y = new Date(Date.now());
```

I got the same result.

File: src/graphql/resolvers/queries/communities.js (hunk 2 above)
Comment (csell5) [reply src/graphql/resolvers/queries/communities.js:101]: I gotcha..

File: src/lib/slack/callSlackHook.js (hunk 5 above)
Comment (csell5) [reply src/lib/slack/callSlackHook.js:20]: Sentry seems like a good idea.

File: src/graphql/resolvers/queries/communities.js (hunk 3 above)
Comment (brettski) [reply src/graphql/resolvers/queries/communities.js:86]: Yeah, my error, I wasn't in the REPL

Comment (csell5) [reply src/graphql/resolvers/queries/communities.js:86]: new dates are hard. common core.

File: src/graphql/resolvers/queries/event.js (hunk 1 above)
Comment (brettski) [reply src/graphql/resolvers/queries/event.js]: fixed in commit 05d7e4d

File: src/graphql/resolvers/queries/communities.js (hunk 3 above)
Comment (brettski) [reply src/graphql/resolvers/queries/communities.js]: fixed in commit 05d7e4d

File: src/lib/slack/callSlackHook.js (hunk 5 above)
Comment (brettski) [reply src/lib/slack/callSlackHook.js:21]: Corrected in commit 761deac

**Comment by brettski** [comment]
@csell5 comment above corrected

**Review by csell5 — APPROVED** [review]

**Pull request merged**